hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMBucketLayoutUpgrade.java

```diff
@@ -0,0 +1,263 @@
+import org.apache.hadoop.ozone.MiniOzoneHAClusterImpl;
+import org.apache.hadoop.ozone.client.ObjectStore;
+import org.apache.hadoop.ozone.client.OzoneClientFactory;
+import org.apache.hadoop.ozone.client.protocol.ClientProtocol;
+import org.apache.hadoop.ozone.om.exceptions.OMException;
+import org.apache.hadoop.ozone.om.helpers.BucketLayout;
+import org.apache.hadoop.ozone.om.helpers.OmBucketInfo;
+import org.apache.hadoop.ozone.om.helpers.OmVolumeArgs;
+import org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol;
+import org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature;
+import org.apache.hadoop.ozone.upgrade.UpgradeFinalizer;
+import org.apache.ozone.test.LambdaTestUtils;
+import org.junit.Test;
+import org.junit.Before;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.UUID;
+
+import static org.apache.hadoop.ozone.OzoneConsts.LAYOUT_VERSION_KEY;
+import static org.apache.hadoop.ozone.om.OMUpgradeTestUtils.waitForFinalization;
+import static org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature.ERASURE_CODED_STORAGE_SUPPORT;
+import static org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature.INITIAL_VERSION;
+import static org.apache.hadoop.ozone.om.upgrade.OMLayoutVersionManager.maxLayoutVersion;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Upgrade testing for Bucket Layout Feature.
+ * <p>
+ * Expected behavior:
+ * 1. Pre-Finalize: OM should not allow creation of buckets with new bucket
+ * layouts. Only LEGACY buckets are allowed.
+ * <p>
+ * 2. Post-Finalize: OM should allow creation of buckets with new bucket
+ * layouts.
+ */
+@RunWith(Parameterized.class)
+public class TestOMBucketLayoutUpgrade {
+  /**
+   * Set a timeout for each test.
+   */
+  @Rule
+  public Timeout timeout = new Timeout(300000);
+  private MiniOzoneHAClusterImpl cluster;
+  private OzoneManager ozoneManager;
+  private ClientProtocol clientProtocol;
+  private static final BucketLayout[] BUCKET_LAYOUTS =
+      new BucketLayout[]{
+          BucketLayout.FILE_SYSTEM_OPTIMIZED,
+          BucketLayout.OBJECT_STORE,
+          BucketLayout.LEGACY
+      };
+  private static final String VOLUME_NAME = "vol-" + UUID.randomUUID();
+  private int fromLayoutVersion;
+  private OzoneManagerProtocol omClient;
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(TestOMBucketLayoutUpgrade.class);
+
+  /**
+   * Defines a "from" layout version to finalize from.
+   *
+   * @return
+   */
+  @Parameterized.Parameters
+  public static Collection<Object[]> data() {
+    return Arrays.asList(new Object[][]{
+        {INITIAL_VERSION},
+        {ERASURE_CODED_STORAGE_SUPPORT},
```

Comment:
We only need to run this suite with the older versions of the system. Since our main focus is to test pre-finalization behaviour.
The below line of code verifies the same.
```java
org.junit.Assume.assumeTrue("Check if there is need to finalize.",
        maxLayoutVersion() > fromLayoutVersion);
```

There would be no need to test for pre-finalization behaviour if we were already at the latest version of the system (`BUCKET_LAYOUT_SUPPORT`) - since there would be nothing to upgrade to.

Reply:
I think we would be ok just to test finalizing from INITIAL_VERSION. Testing ERASURE_CODED_STORAGE_SUPPORT -> BUCKET_LAYOUT_SUPPORT finalization is a subset of what is tested in INITIAL_VERSION -> BUCKET_LAYOUT_SUPPORT finalization.

Reply:
Thank you for clarifying @errose28, will make this change.